File: ganilib-core/src/main/java/com/gani/lib/ui/layout/GRelativeLayout.java
package com.gani.lib.ui.layout;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.widget.RelativeLayout;

import com.gani.lib.ui.view.ViewHelper;

public class GRelativeLayout<T extends GRelativeLayout> extends RelativeLayout {
  private ViewHelper helper;

  public GRelativeLayout(Context context) {
    super(context);

    init();
  }

  private void init() {
    this.helper = new ViewHelper(this);
  }

  private T self() {
    return (T) this;
  }

  public T width(Integer width) {
    helper.width(width);
    return self();
  }

  public T height(Integer height) {
    helper.height(height);
    return self();
  }

  public T padding(Integer left, Integer top, Integer right, Integer bottom) {
    helper.padding(left, top, right, bottom);
    return self();
  }

  public T margin(Integer left, Integer top, Integer right, Integer bottom) {
    helper.margin(left, top, right, bottom);
    return self();
  }

  public T gravity(int gravity) {
    setGravity(gravity);
    return self();
  }

  public T bgColor(int color) {
    helper.bgColor(color);
    return self();
  }

  public T bgColor(String color) {
    helper.bgColor(color);
    return self();
  }

  public T bg(Drawable drawable) {
    helper.bg(drawable);
    return self();
  }

  public T append(View child) {
    addView(child);
    return self();
  }
}

File: ganilib-core/src/main/java/com/gani/lib/ui/layout/GRelativeLayout.kt
package com.gani.lib.ui.layout

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import android.widget.RelativeLayout
import com.gani.lib.ui.view.IView

import com.gani.lib.ui.view.ViewHelper

open class GRelativeLayout<T : GRelativeLayout<T>>: RelativeLayout, IView {
    private val helper: ViewHelper = ViewHelper(this)

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    private fun init() {
        // Nothing to do
    }

    private fun self(): T {
        return this as T
    }

    override fun width(width: Int?): T {
        helper.width(width)
        return self()
    }

    override fun height(height: Int?): T {
        helper.height(height)
        return self()
    }

    override fun padding(l: Int?, t: Int?, r: Int?, b: Int?): T {
        helper.padding(l, t, r, b)
        return self()
    }

    override fun margin(l: Int?, t: Int?, r: Int?, b: Int?): T {
        helper.margin(l, t, r, b)
        return self()
    }

    override fun bgColor(color: Int): T {
        helper.bgColor(color)
        return self()
    }

    fun bgColor(color: String): T {
        helper.bgColor(color)
        return self()
    }

    fun gravity(gravity: Int): T {
        setGravity(gravity)
        return self()
    }

    fun bg(drawable: Drawable): T {
        helper.bg(drawable)
        return self()
    }

    fun append(child: View): T {
        addView(child)
        return self()
    }

    fun relative(addRules: (GRelativeLayoutParams) -> (Unit)): T {
        helper.relative(addRules)
        return self()
    }
}
